src/registrar/tests/services/test_dns_host_service.py

```diff
@@ -409,3 +432,129 @@ def test_save_db_zone_on_failed_join_creation_throws_error(self):
+            {"test_name": "Empty result dictionary test case", "result": {}},
+            {
+                "test_name": "Missing id test case",
+                "result": {
+                    "type": "A",
+                    "name": "test.gov",  # record name
+                    "content": "1.1.1.1",  # IPv4
+                    "ttl": 1,
+                    "comment": "Test record",
+                    "created_on": "2024-01-02T03:04:05Z",
+                    "tags": [],
+                },
+            },
+            {
+                "test_name": "Missing name test case",
+                "result": {
+                    "id": "1234",
+                    "type": "A",
+                    "content": "1.1.1.1",  # IPv4
+                    "ttl": 1,
+                    "comment": "Test record",
+                    "created_on": "2024-01-02T03:04:05Z",
+                    "tags": [],
+                },
+            },
+        ]
+        x_zone_id = self.vendor_account_data["result"].get("id")
+        domain_name = "dns-test.gov"
+        Domain.objects.create(name=domain_name)
+
+        # Create account and zone associated with record
+        self.service.save_db_account(self.vendor_account_data)
+        self.service.save_db_zone(self.vendor_zone_data, domain_name)
+
+        expected_vendor_records = VendorDnsRecord.objects.count()
+        expected_dns_records = DnsRecord.objects.count()
+        expected_record_joins = RecordsJoin.objects.count()
+
+        for payload in invalid_result_payloads:
+            with self.subTest(msg=payload["test_name"], payload=payload):
+                with self.assertRaises(KeyError):
+                    self.service.save_db_record(x_zone_id, payload)
+
+                    # Nothing should be written on any failure
+                    self.assertEqual(VendorDnsAccount.objects.count(), expected_vendor_records)
+                    self.assertEqual(DnsAccount.objects.count(), expected_dns_records)
+                    self.assertEqual(AccountsJoin.objects.count(), expected_record_joins)
+
+    def test_save_db_record_on_failed_join_creation_throws_error(self):
+        x_zone_id = self.vendor_account_data["result"].get("id")
+        domain_name = "dns-test.gov"
+        Domain.objects.create(name=domain_name)
+
+        # Create account and zone associated with record
+        self.service.save_db_account(self.vendor_account_data)
+        self.service.save_db_zone(self.vendor_zone_data, domain_name)
+
+        with patch(
+            "registrar.models.DnsRecord_VendorDnsRecord.objects.create",
+            side_effect=IntegrityError("simulated join failure"),
+        ):
+            with self.assertRaises(IntegrityError):
+                self.service.save_db_record(x_zone_id, self.vendor_record_data)
+
+        # If the creation of the join fails, nothing should be saved in the database.
+        self.assertEqual(VendorDnsRecord.objects.count(), 0)
```

Comment:
Rather than assert the total count in the db, this test would be more robust if you check the total count before running the method you're testing, and then comparing it to the count after (like you are doing in the previous test)